File: src/mirrorr/api/schemas.py
"""Pydantic request/response schemas for API endpoints.

These replace the raw `dict[str, Any] = Body(...)` pattern with proper
Pydantic models, enabling:
1. Automatic OpenAPI schema generation
2. Request validation before hitting the endpoint logic
3. Clearer API contracts for frontend integration
"""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


# ── Auth ───────────────────────────────────────────────────────────────


class LoginRequest(BaseModel):
    username: str = Field(max_length=64)
    password: str = Field(max_length=128)


class RegisterRequest(BaseModel):
    username: str = Field(max_length=64)
    password: str = Field(min_length=8, max_length=128)
    display_name: str = Field(default="", max_length=128)


class RefreshRequest(BaseModel):
    refresh_token: str


class ChangePasswordRequest(BaseModel):
    old_password: str = Field(max_length=128)
    new_password: str = Field(min_length=8, max_length=128)


# ── Sessions ───────────────────────────────────────────────────────────


class CreateSessionRequest(BaseModel):
    profile_id: int | None = None
    engine_id: int
    resolver_id: int
    resolver_config: dict[str, Any] = Field(default_factory=dict)
    retry_mode: str = "none"
    retry_config: dict[str, Any] = Field(default_factory=dict)
    recording: bool = False


# ── Autoruns ───────────────────────────────────────────────────────────


class CreateAutorunRequest(BaseModel):
    user_friendly_name: str
    snake_case_name: str = Field(pattern=r'^[a-zA-Z0-9_-]+$')
    profile_id: int | None = None
    engine_id: int
    resolver_id: int
    resolver_config: dict[str, Any] = Field(default_factory=dict)
    retry_mode: str = "none"
    retry_config: dict[str, Any] = Field(default_factory=dict)
    start_time: datetime
    end_time: datetime
    recording: bool = True


class UpdateAutorunRequest(BaseModel):
    user_friendly_name: str | None = None
    snake_case_name: str | None = Field(default=None, pattern=r'^[a-zA-Z0-9_-]+$')
    profile_id: int | None = None
    engine_id: int | None = None
    resolver_id: int | None = None
    resolver_config: dict[str, Any] | None = None
    retry_mode: str | None = None
    retry_config: dict[str, Any] | None = None
    start_time: datetime | None = None
    end_time: datetime | None = None
    recording: bool | None = None


# ── Profiles ───────────────────────────────────────────────────────────


class CreateProfileRequest(BaseModel):
    name: str
    default_engine_id: int
    resolver_id: int
    resolver_config: dict[str, Any] = Field(default_factory=dict)
    retry_mode: str = "none"
    retry_config: dict[str, Any] = Field(default_factory=dict)


class UpdateProfileRequest(BaseModel):
    name: str | None = None
    default_engine_id: int | None = None
    resolver_id: int | None = None
    resolver_config: dict[str, Any] | None = None
    retry_mode: str | None = None
    retry_config: dict[str, Any] | None = None


class SaveProfileRequest(BaseModel):
    name: str


class CreateClientRequest(BaseModel):
    name: str = Field(max_length=128)


# ── Response Models ────────────────────────────────────────────────────


class UserResponse(BaseModel):
    id: int
    username: str
    role: str
    display_name: str


class ClientResponse(BaseModel):
    id: int
    name: str


class AuthResponse(BaseModel):
    user: UserResponse
    client: ClientResponse | None = None


class SessionResponse(BaseModel):
    id: int
    profile_id: int | None = None
    autorun_id: int | None = None
    engine_id: int
    resolver_id: int
    resolver_config: dict[str, Any] = Field(default_factory=dict)
    retry_mode: str = "none"
    retry_config: dict[str, Any] = Field(default_factory=dict)
    status: str
    recording: bool = False
    retry_attempts: int = 0
    started_at: datetime | None = None
    ended_at: datetime | None = None
    requester_user_token: str = ""
    session_urls: list[dict[str, str]] = Field(default_factory=list)
    attempts: list[dict[str, Any]] = Field(default_factory=list)
    engine_name: str | None = None
    resolver_name: str | None = None
    profile_name: str | None = None
    recording_progress: dict[str, Any] | None = None
    session_folder: str | None = None


class AutorunResponse(BaseModel):
    id: int
    user_friendly_name: str
    snake_case_name: str
    profile_id: int | None = None
    engine_id: int
    resolver_id: int
    resolver_config: dict[str, Any] = Field(default_factory=dict)
    retry_mode: str = "none"
    retry_config: dict[str, Any] = Field(default_factory=dict)
    status: str
    start_time: datetime
    end_time: datetime
    recording: bool = True
    requester_user_token: str = ""
    engine_name: str | None = None
    resolver_name: str | None = None
    profile_name: str | None = None
    next_run_at: datetime | None = None
    last_run_at: datetime | None = None


class RecordingResponse(BaseModel):
    id: int
    user_friendly_name: str
    snake_case_name: str
    disk_path: str
    content_url: str
    profile_name: str
    engine_name: str
    resolver_name: str
    started_at: datetime
    ended_at: datetime
    duration_seconds: float
    size_bytes: int
    created_at: datetime
    requester_user_token: str = ""
    session_id: int | None = None
    media_served: bool = False


class ProfileResponse(BaseModel):
    id: int
    name: str
    default_engine_id: int
    resolver_id: int
    resolver_config: dict[str, Any] = Field(default_factory=dict)
    retry_mode: str = "none"
    retry_config: dict[str, Any] = Field(default_factory=dict)
    requester_user_token: str = ""
    engine_name: str | None = None
    resolver_name: str | None = None


class EngineResponse(BaseModel):
    id: int
    name: str
    description: str = ""
    origin: str
    origin_hash: str
    capabilities: dict[str, Any] = Field(default_factory=dict)
    retry_modes_schema: dict[str, Any] = Field(default_factory=dict)


class ResolverResponse(BaseModel):
    id: int
    name: str
    description: str
    origin: str
    origin_hash: str
    config_schema: dict[str, Any] = Field(default_factory=dict)


class PaginatedResponse(BaseModel):
    items: list[Any]
    next_cursor: int | None = None
    has_more: bool = False
    total: int | None = None


# ── Typed list responses (for OpenAPI schema + runtime validation) ────
# Generic PaginatedResponse uses `list[Any]`, which produces no schema for
# the item shape. These typed variants give the frontend a real contract.


class SessionListResponse(BaseModel):
    items: list[SessionResponse]
    next_cursor: int | None = None
    has_more: bool = False
    total: int | None = None


class AutorunListResponse(BaseModel):
    items: list[AutorunResponse]
    next_cursor: int | None = None
    has_more: bool = False
    total: int | None = None


class RecordingListResponse(BaseModel):
    items: list[RecordingResponse]
    next_cursor: int | None = None
    has_more: bool = False
    total: int | None = None


class ProfileListResponse(BaseModel):
    items: list[ProfileResponse]
    next_cursor: int | None = None
    has_more: bool = False
    total: int | None = None


class EngineListResponse(BaseModel):
    items: list[EngineResponse]
    next_cursor: int | None = None
    has_more: bool = False
    total: int | None = None


class ResolverListResponse(BaseModel):
    items: list[ResolverResponse]
    next_cursor: int | None = None
    has_more: bool = False
    total: int | None = None


# ── Logs / progress / delete-result responses ─────────────────────────


class SessionLogsResponse(BaseModel):
    """Tail of a session's process output log."""

    session_id: int
    stream: str
    name: str = ""
    lines: list[str] = Field(default_factory=list)
    next_offset: int = 0
    eof: bool = True


class RecordingProgressResponse(BaseModel):
    """Latest remux progress (same shape as the NATS subject payload)."""

    session_id: int
    progress: dict[str, Any] | None = None
    status: str = ""


class DeleteResultResponse(BaseModel):
    """Result of a resource deletion, including cascade counts."""

    status: str = "deleted"
    deleted: dict[str, int] = Field(default_factory=dict)

File: mirrorr-web-legacy/src/lib/schemas.ts
import { z } from "zod";

// ── Auth ───────────────────────────────────────────────────────────

export const loginSchema = z.object({
	username: z.string().min(1, "Username is required"),
	password: z.string().min(8, "Password must be at least 8 characters"),
});

export const registerSchema = z.object({
	username: z.string().min(1, "Username is required"),
	password: z.string().min(8, "Password must be at least 8 characters"),
	display_name: z.string().optional(),
});

export const changePasswordSchema = z
	.object({
		old_password: z.string().min(1, "Current password is required"),
		new_password: z.string().min(8, "Password must be at least 8 characters"),
		confirm_password: z
			.string()
			.min(8, "Password must be at least 8 characters"),
	})
	.refine((data) => data.new_password === data.confirm_password, {
		message: "Passwords do not match",
		path: ["confirm_password"],
	});

// ── User ───────────────────────────────────────────────────────────

export const userSchema = z.object({
	id: z.number(),
	username: z.string(),
	role: z.enum(["admin", "user"]),
	display_name: z.string(),
	created_at: z.string().optional(),
});

export type User = z.infer<typeof userSchema>;

// ── Auth Status ────────────────────────────────────────────────────

export const authStatusSchema = z.object({
	has_users: z.boolean(),
	version: z.string().optional(),
	ready: z.boolean().optional(),
});

export const meSchema = z.object({
	user: userSchema,
	client: z.object({ id: z.number(), name: z.string() }).nullable(),
	access_token: z.string().nullable().optional(),
	refresh_token: z.string().nullable().optional(),
});

export const registrationPendingSchema = z.object({
	status: z.literal("pending"),
	username: z.string(),
});
export type RegistrationPending = z.infer<typeof registrationPendingSchema>;

/** Auth response shape returned by /auth/login, /auth/refresh, /auth/me */
export const authResponseSchema = meSchema;
export type AuthResponse = z.infer<typeof authResponseSchema>;

/** Register may return pending instead of a live user (non-first-user path → 202). */
export const registerResponseSchema = z.union([meSchema, registrationPendingSchema]);
export type RegisterResponse = z.infer<typeof registerResponseSchema>;

// ── Registration Requests (admin queue) ────────────────────────────

export const registrationRequestSchema = z.object({
	id: z.number(),
	username: z.string(),
	display_name: z.string(),
	created_at: z.string().nullable(),
});

export type RegistrationRequest = z.infer<typeof registrationRequestSchema>;

// ── Engine ─────────────────────────────────────────────────────────

export const engineCapabilitiesSchema = z.object({
	can_record: z.boolean().default(false),
	can_playlist: z.boolean().default(false),
});

export const engineSchema = z.object({
	id: z.number(),
	name: z.string(),
	description: z.string(),
	origin: z.string(),
	origin_hash: z.string(),
	capabilities: engineCapabilitiesSchema,
	retry_modes_schema: z.record(z.string(), z.any()),
});

export type Engine = z.infer<typeof engineSchema>;

// ── Resolver ───────────────────────────────────────────────────────

export const resolverSchema = z.object({
	id: z.number(),
	name: z.string(),
	description: z.string(),
	origin: z.string(),
	origin_hash: z.string(),
	config_schema: z.record(z.string(), z.any()),
});

export type Resolver = z.infer<typeof resolverSchema>;

// ── Profile ────────────────────────────────────────────────────────

export const profileSchema = z.object({
	id: z.number(),
	name: z.string(),
	default_engine_id: z.number(),
	resolver_id: z.number(),
	resolver_config: z.record(z.string(), z.any()),
	retry_mode: z.string(),
	retry_config: z.record(z.string(), z.any()),
	requester_user_token: z.string().optional(),
	engine_name: z.string().nullable().optional(),
	resolver_name: z.string().nullable().optional(),
});

export type Profile = z.infer<typeof profileSchema>;

export const createProfileSchema = z.object({
	name: z.string().min(1, "Name is required"),
	default_engine_id: z.number({ message: "Engine is required" }),
	resolver_id: z.number({ message: "Resolver is required" }),
	resolver_config: z.record(z.string(), z.any()).default({}),
	retry_mode: z.string().default("none"),
	retry_config: z.record(z.string(), z.any()).default({}),
});

// ── Session ────────────────────────────────────────────────────────

export const sessionStatusSchema = z.enum([
	"active",
	"recording",
	"terminating",
	"remuxing",
	"finalizing",
	"completed",
	"failed",
]);

const recordingProgressSchema = z.object({
	percent: z.number(),
	frame: z.number().optional(),
	current_time: z.number().optional(),
	total_duration: z.number().optional(),
	speed: z.number().optional(),
	elapsed: z.number().optional(),
	eta_seconds: z.number().nullable().optional(),
});

const attemptSchema = z.object({
	index: z.number(),
	started_at: z.string().optional(),
	ended_at: z.string().optional(),
	duration_seconds: z.number().optional(),
	returncode: z.number().nullable().optional(),
	reason: z.string().nullable().optional(),
});

export const sessionSchema = z.object({
	id: z.number(),
	profile_id: z.number().nullable(),
	autorun_id: z.number().nullable(),
	engine_id: z.number(),
	resolver_id: z.number(),
	resolver_config: z.record(z.string(), z.any()),
	retry_mode: z.string(),
	retry_config: z.record(z.string(), z.any()),
	status: sessionStatusSchema,
	recording: z.boolean(),
	retry_attempts: z.number(),
	started_at: z.string().nullable(),
	ended_at: z.string().nullable(),
	requester_user_token: z.string().optional(),
	session_urls: z.array(z.object({ label: z.string(), url: z.string() })),
	attempts: z.array(attemptSchema).default([]),
	engine_name: z.string().nullable().optional(),
	resolver_name: z.string().nullable().optional(),
	profile_name: z.string().nullable().optional(),
	recording_progress: recordingProgressSchema.optional(),
	session_folder: z.string().nullable().optional(),
});

export type Session = z.infer<typeof sessionSchema>;

export const createSessionSchema = z.object({
	profile_id: z.number().optional(),
	engine_id: z.number({ message: "Engine is required" }),
	resolver_id: z.number({ message: "Resolver is required" }),
	resolver_config: z.record(z.string(), z.any()).default({}),
	retry_mode: z.string().default("none"),
	retry_config: z.record(z.string(), z.any()).default({}),
	recording: z.boolean().default(false),
});

// ── Autorun ────────────────────────────────────────────────────────

export const autorunSchema = z.object({
	id: z.number(),
	user_friendly_name: z.string(),
	snake_case_name: z.string(),
	profile_id: z.number().nullable(),
	engine_id: z.number(),
	resolver_id: z.number(),
	resolver_config: z.record(z.string(), z.any()),
	retry_mode: z.string(),
	retry_config: z.record(z.string(), z.any()),
	status: z
		.enum([
			"scheduled",
			"active",
			"recording",
			"terminating",
			"remuxing",
			"finalizing",
			"completed",
			"failed",
		])
		.default("scheduled"),
	start_time: z.string(),
	end_time: z.string(),
	recording: z.boolean(),
	requester_user_token: z.string().optional(),
	engine_name: z.string().nullable().optional(),
	resolver_name: z.string().nullable().optional(),
	profile_name: z.string().nullable().optional(),
	next_run_at: z.string().nullable().optional(),
	last_run_at: z.string().nullable().optional(),
});

export type Autorun = z.infer<typeof autorunSchema>;

export const createAutorunSchema = z.object({
	user_friendly_name: z.string().min(1, "Name is required"),
	snake_case_name: z.string().min(1, "Slug is required"),
	profile_id: z.number().optional(),
	engine_id: z.number({ message: "Engine is required" }),
	resolver_id: z.number({ message: "Resolver is required" }),
	resolver_config: z.record(z.string(), z.any()).default({}),
	retry_mode: z.string().default("none"),
	retry_config: z.record(z.string(), z.any()).default({}),
	start_time: z.string().min(1, "Start time is required"),
	end_time: z.string().min(1, "End time is required"),
	recording: z.boolean().default(true),
});

// ── Recording ──────────────────────────────────────────────────────

export const recordingSchema = z.object({
	id: z.number(),
	user_friendly_name: z.string(),
	snake_case_name: z.string(),
	disk_path: z.string(),
	content_url: z.string(),
	profile_name: z.string(),
	engine_name: z.string(),
	resolver_name: z.string(),
	started_at: z.string(),
	ended_at: z.string(),
	duration_seconds: z.number(),
	size_bytes: z.number(),
	created_at: z.string(),
	requester_user_token: z.string().optional(),
	session_id: z.number().nullable().optional(),
	media_served: z.boolean().optional(),
});

export type Recording = z.infer<typeof recordingSchema>;

// ── Notification ───────────────────────────────────────────────────

export const notificationSchema = z.object({
	id: z.number().optional(),
	resource_type: z.string(),
	resource_id: z.number(),
	event_type: z.string(),
	title: z.string(),
	body: z.string().optional(),
	read: z.boolean().optional(),
	created_at: z.string(),
});

export type Notification = z.infer<typeof notificationSchema>;

// ── Session Control ────────────────────────────────────────────────

export const controlResponseSchema = z.object({
	ok: z.literal(true),
	command: z.string(),
});

export type ControlResponse = z.infer<typeof controlResponseSchema>;

// ── Import/Export ─────────────────────────────────────────────────

export const importBundleSchema = z.object({
	profiles: z.array(z.record(z.string(), z.unknown())).optional(),
	autoruns: z.array(z.record(z.string(), z.unknown())).optional(),
});

export type ImportBundle = z.infer<typeof importBundleSchema>;

export const validationReportSchema = z.object({
	valid: z.boolean(),
	profiles: z.array(
		z.object({
			name: z.string(),
			content_hash: z.string(),
			valid: z.boolean(),
			issues: z.array(
				z.object({
					field: z.string().optional(),
					problem: z.string(),
					bundled: z
						.object({ name: z.string(), origin_hash: z.string() })
						.optional(),
					installed: z
						.object({ name: z.string(), origin_hash: z.string() })
						.optional(),
					alternatives: z
						.array(
							z.object({
								id: z.number(),
								name: z.string(),
								origin_hash: z.string(),
							}),
						)
						.optional(),
					existing_id: z.number().optional(),
					profile_name: z.string().optional(),
					new_name: z.string().optional(),
				}),
			),
		}),
	),
	autoruns: z.array(
		z.object({
			name: z.string(),
			valid: z.boolean(),
			issues: z.array(
				z.object({
					field: z.string().optional(),
					problem: z.string(),
					bundled: z
						.object({ name: z.string(), origin_hash: z.string() })
						.optional(),
					installed: z
						.object({ name: z.string(), origin_hash: z.string() })
						.optional(),
					alternatives: z
						.array(
							z.object({
								id: z.number(),
								name: z.string(),
								origin_hash: z.string(),
							}),
						)
						.optional(),
					profile_name: z.string().optional(),
				}),
			),
		}),
	),
});

export type ValidationReport = z.infer<typeof validationReportSchema>;

// ── Update Autorun (validated request body) ──────────────────────

export const updateAutorunSchema = z.object({
	user_friendly_name: z.string().min(1).optional(),
	snake_case_name: z
		.string()
		.regex(/^[a-zA-Z0-9_-]+$/)
		.optional(),
	profile_id: z.number().optional(),
	engine_id: z.number().optional(),
	resolver_id: z.number().optional(),
	resolver_config: z.record(z.string(), z.any()).optional(),
	retry_mode: z.string().optional(),
	retry_config: z.record(z.string(), z.any()).optional(),
	start_time: z.string().optional(),
	end_time: z.string().optional(),
	recording: z.boolean().optional(),
});

export type UpdateAutorunPayload = z.infer<typeof updateAutorunSchema>;

// ── Import/Export apply response ─────────────────────────────────

export const applyResponseSchema = z.object({
	profiles_created: z.number(),
	autoruns_created: z.number(),
	profiles_skipped: z.number(),
	created: z
		.object({
			profiles: z
				.array(z.object({ id: z.number(), name: z.string() }))
				.optional(),
			autoruns: z
				.array(z.object({ id: z.number(), name: z.string() }))
				.optional(),
		})
		.optional(),
	renamed: z.record(z.string(), z.string()).optional(),
});

export type ApplyResponse = z.infer<typeof applyResponseSchema>;

// ── Logs / progress / delete result ───────────────────────────────

export const deleteResultSchema = z.object({
	status: z.string().default("deleted"),
	deleted: z.record(z.string(), z.number()),
});

export type DeleteResult = z.infer<typeof deleteResultSchema>;

export const sessionLogsSchema = z.object({
	session_id: z.number(),
	stream: z.string(),
	name: z.string().default(""),
	lines: z.array(z.string()).default([]),
	next_offset: z.number().default(0),
	eof: z.boolean().default(true),
});

export type SessionLogs = z.infer<typeof sessionLogsSchema>;

export const recordingProgressResponseSchema = z.object({
	session_id: z.number(),
	progress: recordingProgressSchema.nullable().optional(),
	status: z.string().default(""),
});

export type RecordingProgressResponse = z.infer<typeof recordingProgressResponseSchema>;
